share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -537,38 +539,322 @@ private static ShareGroupOffset merge(ShareGroupOffset soFar, ShareUpdateValue n
+            PersisterStateBatch candidate,
+            List<PersisterStateBatch> nonOverlapping
+        ) {
+            this.last = last;
+            this.candidate = candidate;
+            this.nonOverlapping = nonOverlapping;
+        }
+
+        public PersisterStateBatch last() {
+            return last;
+        }
+
+        public PersisterStateBatch candidate() {
+            return candidate;
+        }
+
+        public List<PersisterStateBatch> nonOverlapping() {
+            return nonOverlapping;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (!(o instanceof BatchOverlapState)) return false;
+            BatchOverlapState that = (BatchOverlapState) o;
+            return Objects.equals(last, that.last) && Objects.equals(candidate, that.candidate) && Objects.equals(nonOverlapping, that.nonOverlapping);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(last, candidate, nonOverlapping);
+        }
+    }
+
+    static List<PersisterStateBatch> mergeBatches(List<PersisterStateBatch> batches) {
+        if (batches.size() < 2) {
+            return batches;
+        }
+        TreeSet<PersisterStateBatch> sortedBatches = new TreeSet<>(batches);
+        List<PersisterStateBatch> finalBatches = new ArrayList<>(batches.size() * 2); // heuristic size
+
+        BatchOverlapState overlapState = getOverlappingState(sortedBatches);
+
+        while (overlapState != BatchOverlapState.SENTINEL) {
+            PersisterStateBatch last = overlapState.last();
+            PersisterStateBatch candidate = overlapState.candidate();
+            
+            // remove non overlapping prefix from sortedBatches,
+            // will make getting next overlapping pair efficient
+            // as a prefix batch which is non overlapping will only
+            // be checked once.
+            if (overlapState.nonOverlapping() != null) {
+                overlapState.nonOverlapping().forEach(sortedBatches::remove);
+                finalBatches.addAll(overlapState.nonOverlapping());
+            }
+
+            if (candidate == null) {
+                overlapState = BatchOverlapState.SENTINEL;
+                continue;
+            }
+
+            // overlap and same state (last.firstOffset <= candidate.firstOffset) due to sort
+            // covers:
+            // case:        1        2          3            4          5           6          7 (contiguous)
+            // last:        ______   _______    _______      _______   _______   ________    _______
+            // candidate:   ______   ____       __________     ___        ____       _______        _______
+            if (compareBatchState(candidate, last) == 0) {
+                sortedBatches.remove(last);  // remove older smaller interval
+                sortedBatches.remove(candidate);
+
+                last = new PersisterStateBatch(
```

Comment:
There's no need to assign to `last` here. The pattern you've followed in the later cases simply adds the constructed object.